### ISSUE TITLE
Incorrect named vector lookup inside mutate

### ISSUE THREAD

**Issue by dhimmel** [hadley/dplyr#1363]
I recently experienced an unexpected and silent R behavior with devastating consequences. I am not sure the issue is caused by dplyr, but figure this community can direct me to the right place if not.

**Description:** Value-lookup using vector/list keys breaks inside `dplyr::mutate`.

After a wild [goose chase](https://github.com/dhimmel/snplentiful/issues/3), I was able to track down the bug. Observe the following in the [notebook documenting the bug](https://github.com/dhimmel/snplentiful/blob/d6f1187a237d7d6588aa51bb6536f381577d71d1/visualization.ipynb):

+ In cell 2, I create a named vector to map codenames to human readable names.
+ In cell 3, I read a tsv into a dataframe.
+ In cell 4, I use `tidyr::gather` to create a single observation per row. I also add the human readable names using `dplyr::mutate(platform_name = platforms[platform])`
+ In cell 5, I show that the human readable names were incorrectly mapped to the codenames. For example `snps_affy500_log` is paired with `Illumina HumanOmni1` rather than `Affymetrix 500K Set`.

Perhaps this occurred due to nasty factors that somehow infiltrated my hadleyverse. In any case, the behavior is highly unexpected and dangerous. Any ideas?

**Comment by romainfrancois** [comment]
Did you try the dev version. mutate has been fixee in dev. 

If the problem persists when you try dev, then please make a reproducible example.

**Comment by dhimmel** [comment]
Okay I tracked down the cause: using a factor for named vector lookup. Since it's not a dplyr bug, installing dplyr from GitHub did not resolve the issue.

I have simplified the [notebook](https://github.com/dhimmel/snplentiful/blob/ec4366969d65096f5251ba394aa534eb7ed068ac/visualization.ipynb) and added additional documentation of the issue. The issue was caused by the following steps:

1. `tidyr::gather` makes the key column a factor
2. named vector lookup with a factor uses the factor's integer rather than character value

So in conclusion this seems like a serious *gotcha* inherent to factors. While it conceptually makes sense for `tidyr::gather` to make the key a factor, the downstream consequences can be unpleasant. Do you suggest posting this to the [tidyr issues](https://github.com/hadley/tidyr/issues) page, or is this a behavior I should just get used to? I assume name vector lookup is too ancient and embedded of a feature to change?

**Comment by dhimmel** [comment]
Hat tip to @a-pankov for the initial suggestion that factors were to blame (via a personal communication).
